lib/core/logger.cpp

```diff
@@ -0,0 +1,108 @@
+#include "logger.hpp"
+
+#include <chrono>
+
+namespace hyped::utils {
+void Logger::set_level(const Level level)
+{
+  level_ = level;
+};
+
+void Logger::intToLevel(const int level)
+{
+  switch (level) {
+    case 0:
+      level_ = Level::kNone;
+      break;
+    case 1:
+      level_ = Level::kDebug;
+      break;
+    case 2:
+      level_ = Level::kInfo;
+      break;
+    case 3:
+      level_ = Level::kWarn;
+      break;
+    case 4:
+      level_ = Level::kFatal;
+      break;
+    default:
+      level_ = Level::kNone;
+      break;
+  }
+};
+
+void Logger::printHead(FILE *file, const char *title)
+{
+  std::time_t t = std::time(nullptr);
+  tm *tt        = localtime(&t);
+  fprintf(file, "%02d:%02d:%02d", tt->tm_hour, tt->tm_min, tt->tm_sec);
+
+  static const bool print_micro = true;
+  if (print_micro) {
+    auto now_time = std::chrono::high_resolution_clock::now().time_since_epoch();
+    std::chrono::duration<int, std::milli> time_span
+      = duration_cast<std::chrono::milliseconds>(now_time);
+    fprintf(file, ".%03d ", static_cast<uint16_t>(time_span.count()) % 1000);
+  } else {
+    fprintf(file, " ");
+  }
+  fprintf(file, "%s[%s]: ", title, module_);
+};
+
+void Logger::print(FILE *file, const char *format, va_list args)
+{
+  vfprintf(file, format, args);
+  fprintf(file, "\n");
+};
+
+void Logger::debug(const char *format, ...)
+{
+  FILE *file = stdout;
+  if (level_ == Level::kDebug) {
+    printHead(file, "DEBUG");
+    va_list args;
```

Comment:
Is there a reason for doing `va_list args`, `va_start` etc. in each logging level instead of within print itself?

Reply:
Can only use `va_start` within a function that accepts `...` (or ellipsis) as a parameter

Reply:
makes sense